### DISCUSSION TITLE
Logstash won't start after upgrading to 2.3.110

### DISCUSSION THREAD

**Discussion by jaek79** [Security-Onion-Solutions/securityonion#7655] [Unsupported Versions] [answered]
Hello,

I have performed the version upgrade to 2.3.110 and started seeing the below issue. 
I could able to start other services but logstash won't start and hang at Wait_start status. 
I have included the below details.
Please if anyone could help I would greatly appreciate it. 

standalone 
2.3.110

error when run sudo salt-call state.highstate:

error response from daemon: No such container: so-logstash
Error: No such container: so-logstash
[ERROR   ] Rendering exception occurred
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/salt/utils/templates.py", line 502, in render_jinja_tmpl
    output = template.render(**decoded_context)
  File "/usr/lib/python3.6/site-packages/jinja2/environment.py", line 1090, in render
    self.environment.handle_exception()
  File "/usr/lib/python3.6/site-packages/jinja2/environment.py", line 832, in handle_exception
    reraise(*rewrite_traceback_stack(source=source))
  File "/usr/lib/python3.6/site-packages/jinja2/_compat.py", line 28, in reraise
    raise value.with_traceback(tb)
  File "<template>", line 39, in top-level template code
  File "/usr/lib/python3.6/site-packages/jinja2/environment.py", line 1155, in make_module
    return TemplateModule(self, self.new_context(vars, shared, locals))
  File "/usr/lib/python3.6/site-packages/jinja2/environment.py", line 1238, in __init__
    body_stream = list(template.root_render_func(context))
  File "/var/cache/salt/minion/files/base/logstash/map.jinja", line 7, in top-level template code
    {% for node_type, node_details in node_data.items() | sort %}
  File "/usr/lib/python3.6/site-packages/jinja2/sandbox.py", line 460, in call
    if not __self.is_safe_callable(__obj):
  File "/usr/lib/python3.6/site-packages/jinja2/sandbox.py", line 360, in is_safe_callable
    getattr(obj, "unsafe_callable", False) or getattr(obj, "alters_data", False)
jinja2.exceptions.UndefinedError: 'None' has no attribute 'items'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/salt/utils/templates.py", line 261, in render_tmpl
    output = render_str(tmplstr, context, tmplpath)
  File "/usr/lib/python3.6/site-packages/salt/utils/templates.py", line 509, in render_jinja_tmpl
    raise SaltRenderError("Jinja variable {}{}".format(exc, out), buf=tmplstr)
salt.exceptions.SaltRenderError: Jinja variable 'None' has no attribute 'items'
/var/cache/salt/minion/files/base/logstash/map.jinja(7):
---
[...]
{% set REDIS_NODES = [] %}
{% set mainint = salt['pillar.get']('host:mainint') %}
{% set localhostip = salt['grains.get']('ip_interfaces').get(mainint)[0] %}
{% if role in ['so-node', 'so-standalone', 'so-managersearch'] %}
  {% set node_data = salt['pillar.get']('logstash:nodes') %}
  {% for node_type, node_details in node_data.items() | sort %}    <======================
    {% if node_type in ['manager', 'managersearch', 'standalone', 'receiver' ] %}
      {% for hostname in node_data[node_type].keys() %}
        {% do REDIS_NODES.append({hostname:node_details[hostname].ip}) %}
      {% endfor %}
    {% endif %}
[...]
---
[CRITICAL] Rendering SLS 'base:logstash' failed: Jinja variable 'None' has no attribute 'items'
/var/cache/salt/minion/files/base/logstash/map.jinja(7):
---
[...]
{% set REDIS_NODES = [] %}
{% set mainint = salt['pillar.get']('host:mainint') %}
{% set localhostip = salt['grains.get']('ip_interfaces').get(mainint)[0] %}
{% if role in ['so-node', 'so-standalone', 'so-managersearch'] %}
  {% set node_data = salt['pillar.get']('logstash:nodes') %}
  {% for node_type, node_details in node_data.items() | sort %}    <======================
    {% if node_type in ['manager', 'managersearch', 'standalone', 'receiver' ] %}
      {% for hostname in node_data[node_type].keys() %}
        {% do REDIS_NODES.append({hostname:node_details[hostname].ip}) %}
      {% endfor %}
    {% endif %}
[...]
---
local:
    Data failed to compile:
----------
    Rendering SLS 'base:logstash' failed: Jinja variable 'None' has no attribute 'items'
/var/cache/salt/minion/files/base/logstash/map.jinja(7):
---
[...]
{% set REDIS_NODES = [] %}
{% set mainint = salt['pillar.get']('host:mainint') %}
{% set localhostip = salt['grains.get']('ip_interfaces').get(mainint)[0] %}
{% if role in ['so-node', 'so-standalone', 'so-managersearch'] %}
  {% set node_data = salt['pillar.get']('logstash:nodes') %}
  {% for node_type, node_details in node_data.items() | sort %}    <======================
    {% if node_type in ['manager', 'managersearch', 'standalone', 'receiver' ] %}
      {% for hostname in node_data[node_type].keys() %}
        {% do REDIS_NODES.append({hostname:node_details[hostname].ip}) %}
      {% endfor %}
    {% endif %}
[...]
---

**Comment by dougburks** [comment]
Have you checked your Logstash configuration for typos or other errors?

Have you tried restarting Logstash?

Have you tried rebooting?

**Reply by jaek79** [reply]
Hi dougburks,

First of all thanks so much for the reply! 
I have answers to your questions below.
Thanks!

1. Have you checked your Logstash configuration for typos or other errors?
--> I have checked I couldn't any issues with the configuration. I haven't made any changes to the config for a long time and we didn't have any problem until the upgrade to 2.3.110. Also, I have attached the logstash.log.

Config: /opt/so/saltstack/local/pillar/minions/****_standalone.sls (I have hided the name of the host)

llogstash_settings:
  ls_pipeline_batch_size: 125
  ls_input_threads: 1
  lsheap: 1000m
  ls_pipeline_workers: 48

Directory:
├── logstash
├── minions
│   └── ****_standalone.sls


2. Have you tried restarting Logstash?
--> Yes, I have tried restarting the service multiple times. 
sudo so-logstash-restart
sudo salt-call state.highstate

3. Have you tried rebooting?
--> Yes, have rebooted the system 2-3 times. 
[logstash.log](https://github.com/Security-Onion-Solutions/securityonion/files/8383036/logstash.log)

**Comment by defensivedepth** [comment, marked as answer]
@jaek79 Is filebeat running?

Can you run `sudo docker logs so-filebeat` and post any output you receive?

**Reply by m0duspwnens** [reply]
@jaek79 Can you verify that the `salt-minion` service is running on all of the nodes in your grid? If the service is running, run `sudo salt \* mine.update` from your manager/standalone. After that completes, run `sudo salt-call state.highstate -linfo` on the standalone.

**Reply by jaek79** [reply]
@defensivedepth 

Thanks for getting back to me.
I have run that command but no output. However, so-filebeat is running. FYI we also do have separate filebeat container running in this server. 

@m0duspwnens 

Thanks so much! After following your suggestion, the problem is fixed now!

- I have checked Grid but no data 
- I have run sudo salt \* test.ping and the result show true on standalone. 
- I have run salt \"name of ids" mine.update 
- I have run sudo salt-call state.highstate -linfo
- Verified all so service is running ok. 
- I have rebooted the server to make sure and confirmed so service started after the reboot.


Once again @m0duspwnens and @defensivedepth thanks so much for your help to resolve this issue. I greatly appreciate it.
